propagation_ot.go

```diff
@@ -0,0 +1,313 @@
+package basictracer
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+type splitTextPropagator struct {
+	tracer *tracerImpl
+}
+type splitBinaryPropagator struct {
+	tracer *tracerImpl
+}
+type goHTTPPropagator struct {
+	*splitBinaryPropagator
+}
+
+const (
+	fieldNameTraceID = "traceid"
+	fieldNameSpanID  = "spanid"
+	fieldNameSampled = "sampled"
+)
+
+func (p *splitTextPropagator) InjectSpan(
+	sp opentracing.Span,
+	carrier interface{},
+) error {
+	sc, ok := sp.(*spanImpl)
+	if !ok {
+		return opentracing.ErrInvalidSpan
+	}
+	splitTextCarrier, ok := carrier.(*opentracing.SplitTextCarrier)
+	if !ok {
+		return opentracing.ErrInvalidCarrier
+	}
+	if splitTextCarrier.TracerState == nil {
+		splitTextCarrier.TracerState = make(map[string]string, 3 /* see below */)
+	}
+	splitTextCarrier.TracerState[fieldNameTraceID] = strconv.FormatInt(sc.raw.TraceID, 10)
+	splitTextCarrier.TracerState[fieldNameSpanID] = strconv.FormatInt(sc.raw.SpanID, 10)
+	splitTextCarrier.TracerState[fieldNameSampled] = strconv.FormatBool(sc.raw.Sampled)
+
+	sc.Lock()
+	if l := len(sc.raw.Baggage); l > 0 && splitTextCarrier.Baggage == nil {
+		splitTextCarrier.Baggage = make(map[string]string, l)
+	}
+	for k, v := range sc.raw.Baggage {
+		splitTextCarrier.Baggage[k] = v
+	}
+	sc.Unlock()
+	return nil
+}
+
+func (p *splitTextPropagator) JoinTrace(
+	operationName string,
+	carrier interface{},
+) (opentracing.Span, error) {
+	splitTextCarrier, ok := carrier.(*opentracing.SplitTextCarrier)
+	if !ok {
+		return nil, opentracing.ErrInvalidCarrier
+	}
+	requiredFieldCount := 0
+	var traceID, propagatedSpanID int64
+	var sampled bool
+	var err error
+	for k, v := range splitTextCarrier.TracerState {
+		switch strings.ToLower(k) {
+		case fieldNameTraceID:
+			traceID, err = strconv.ParseInt(v, 10, 64)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+		case fieldNameSpanID:
+			propagatedSpanID, err = strconv.ParseInt(v, 10, 64)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+		case fieldNameSampled:
+			sampled, err = strconv.ParseBool(v)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+		default:
+			continue
+		}
+		requiredFieldCount++
+	}
+	const expFieldCount = 3
+	if requiredFieldCount < expFieldCount {
+		if len(splitTextCarrier.TracerState) == 0 {
+			return nil, opentracing.ErrTraceNotFound
+		}
+		return nil, opentracing.ErrTraceCorrupted
+	}
+
+	sp := p.tracer.getSpan()
+	sp.raw = RawSpan{
+		Context: Context{
+			TraceID:      traceID,
+			SpanID:       randomID(),
+			ParentSpanID: propagatedSpanID,
+			Sampled:      sampled,
+		},
+	}
+	sp.raw.Baggage = splitTextCarrier.Baggage
```

Comment:
You're looking at unchanged code - this is the awkward move I talked about in the PR message. I won't touch any of the existing serializers in this PR.